### PR TITLE
feature/rgb++

### PR DESCRIPTION
**Pull request by Daryl-L**
1. rgb++ asset in address tab
2. rgb++ transaction digest prompt window for address page

https://github.com/Magickbase/ckb-explorer-public-issues/issues/566

### CODE REVIEW

File: src/pages/Address/AddressAssetComp.tsx [hunk 1]
```diff
@@ -61,7 +65,15 @@ export const AddressSudtComp = ({ account }: { account: SUDT }) => {
   )
 }
 
-export const AddressSporeComp = ({ account }: { account: Spore }) => {
+export const AddressSporeComp = ({
+  account,
+  isRGBPP,
+  udtLabel,
```
Comment (Keith-CY) [inline src/pages/Address/AddressAssetComp.tsx:71]: It's `spore comp` so the label should be `spore`

File: src/pages/Address/AddressAssetComp.tsx [hunk 2]
```diff
@@ -74,16 +86,25 @@ export const AddressSporeComp = ({ account }: { account: Spore }) => {
...
       icon={{ url: patchMibaoImg(icon), errorHandler: handleNftImgError }}
     />
   )
 }
 
-export const AddressMNFTComp = ({ account }: { account: MNFT }) => {
+export const AddressMNFTComp = ({
+  account,
+  isRGBPP,
+  udtLabel,
```
Comment (Keith-CY) [inline src/pages/Address/AddressAssetComp.tsx:102]: It's `m nft comp` so the label should be 'm nft'

File: src/pages/Address/AddressAssetComp.tsx [hunk 3]
```diff
@@ -119,7 +141,15 @@ export const AddressMNFTComp = ({ account }: { account: MNFT }) => {
   )
 }
 
-export const AddressNRC721Comp = ({ account }: { account: NRC721 }) => {
+export const AddressNRC721Comp = ({
+  account,
+  isRGBPP,
+  udtLabel,
```
Comment (Keith-CY) [inline src/pages/Address/AddressAssetComp.tsx:147]: It's `nrc 721 comp` so the label should be `nrc 721`

File: src/pages/Address/AddressAssetComp.tsx [hunk 4]
```diff
@@ -156,7 +187,15 @@ export const AddressNRC721Comp = ({ account }: { account: NRC721 }) => {
   )
 }
 
-export const AddressCoTAComp = ({ account }: { account: CoTA }) => {
+export const AddressCoTAComp = ({
+  account,
+  isRGBPP,
+  udtLabel,
```
Comment (Keith-CY) [inline src/pages/Address/AddressAssetComp.tsx:193]: It's `cota comp` so the label should be `cota`

File: src/pages/Address/AddressAssetComp.tsx [hunk 5]
```diff
@@ -192,18 +232,27 @@ export const AddressCoTAComp = ({ account }: { account: CoTA }) => {
   )
 }
 
-export const AddressOmigaInscriptionComp = ({ account }: { account: OmigaInscription }) => {
+export const AddressOmigaInscriptionComp = ({
+  account,
+  isRGBPP,
+  udtLabel,
```
Comment (Keith-CY) [inline src/pages/Address/AddressAssetComp.tsx:238]: It's `omiga inscription comp` so the label should be `omiga symbol`

File: .eslintrc.js [hunk 6]
```diff
@@ -126,6 +126,7 @@ module.exports = {
       node: {
         paths: ['src'],
         extensions: ['.js', '.jsx', '.ts', '.tsx'],
+        moduleDirectory: ['node_modules', 'src/'],
```
Comment (Keith-CY) [inline .eslintrc.js:129]: What's this update for

File: src/components/TransactionItem/TransactionRGBPPDigestModal/TransactionRGBPPDigestTransfer.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import { TransactionRGBPPDigestTransferAsset } from './TransactionRGBPPDigestTransferAsset'
+import { TransactionRecord } from '../../../services/ExplorerService'
+import AddressText from '../../AddressText'
+
+export const TransactionRGBPPDigestTransfer = ({ transfer }: { transfer: TransactionRecord }) => {
+  const address = parseBTCAddress(transfer.address)
+
+  let addressType = ''
+  if (address) {
+    addressType = address.isBech32 ? `Bech32 (${address.type})` : `Base58 (${address.type})`
```
Comment (Keith-CY) [inline src/components/TransactionItem/TransactionRGBPPDigestModal/TransactionRGBPPDigestTransfer.tsx:12]: The condition is `encoding` is `bech32` or `not bech32` but the else branch treats `not bech32` as `is base58`, which is not proper. `encoding` will fallback to `bech58` when new `encoding` is introduced while this code is not updated.

File: src/components/TransactionItem/TransactionRGBPPDigestModal/TransactionRGBPPDigestContent.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                error => {
+                  console.error(error)
+                },
+              )
+            }}
+          />
+        </div>
+      </div>
+      {(tx.transfers ?? []).map(transfer => (
+        <TransactionRGBPPDigestTransfer transfer={transfer} />
```
Comment (Keith-CY) [inline src/components/TransactionItem/TransactionRGBPPDigestModal/TransactionRGBPPDigestContent.tsx:54]: `No records` should be displayed when the list is empty

File: src/components/TransactionItem/TransactionRGBPPDigestModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  const { data } = useQuery(['rgb-digest', hash], () => explorerService.api.fetchRGBDigest(hash))
+
+  return (
+    <div className={styles.container}>
+      <div className={styles.header}>
+        <span className={styles.left}>{t('address.transaction_rgbpp_digest')}</span>
+        <CloseIcon className={styles.right} onClick={() => onClickClose()} />
+      </div>
+      {data && <TransactionRGBPPDigestContent tx={data.data} leapDirection={leapDirection} />}
```
Comment (Keith-CY) [inline src/components/TransactionItem/TransactionRGBPPDigestModal/index.tsx:28]: Loading should be displayed

File: src/components/TransactionItem/index.tsx [hunk 10]
```diff
@@ -54,6 +62,26 @@ const Time: React.FC<{ tx?: Transaction }> = ({ tx }) => {
   return null
 }
 
+const computeRGBPPCellAmount = (cells: Cell[]) => {
+  return cells.reduce((cur, cell) => {
+    try {
+      const script = addressToScript(cell.addressHash)
+      if (!script) {
```
Comment (Keith-CY) [inline src/components/TransactionItem/index.tsx:69]: In what case will `script` be `falsy`

File: src/components/TransactionItem/index.tsx [hunk 11]
```diff
@@ -54,6 +62,26 @@ const Time: React.FC<{ tx?: Transaction }> = ({ tx }) => {
   return null
 }
 
+const computeRGBPPCellAmount = (cells: Cell[]) => {
+  return cells.reduce((cur, cell) => {
```
Comment (Keith-CY) [inline src/components/TransactionItem/index.tsx:66]: This logic can be done by `filter` as 
```js
cells.filter(c => {
  const script = addressToScript(cell.addressHash)
  cosnt tag = matchScript(script.codeHash, script.hashType)
  return tag?.tag === 'rgb++'
}).length
```

**Comment by Keith-CY** [comment]
Please specify preview link to an address that shows the RBG++ feature in the PR message so we can verify it quickly

**Comment by Keith-CY** [comment]
image
The label of BTC address type is inproper, and the `address` is wrapped unexpectedly

**Comment by Keith-CY** [comment]
Previews at https://ckb-explorer-frontend-in-magickbase-repo-git-7cb256-magickbase.vercel.app/address/16pB1NmhoJaDUVPCc8nmMMtijT31FvGQyj

**Comment by Keith-CY** [comment]
image
The second transfer of `1UJHeZkKA7ieCuDGV3YqYKQnop4uWGKBs` is `spore` but displayed as `CKB`

File: src/components/TransactionItem/index.tsx (hunk 11 above)
Comment (Daryl-L) [reply src/components/TransactionItem/index.tsx]: done

File: src/components/TransactionItem/TransactionRGBPPDigestModal/Amount.tsx [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  } else if (diffStatus === 'positive') {
+    className = styles.increase
+  }
+
+  return (
+    <div className={className}>
+      {brackets && '('}
+      <span className={styles.integer}>{integer}</span>.
+      <span className={styles.decimal}>
+        {decimal ?? '00000000'}
```
[the remark this answers is not in the document]
Comment (Daryl-L) [reply src/components/TransactionItem/TransactionRGBPPDigestModal/Amount.tsx]: done

File: src/components/TransactionItem/TransactionRGBPPDigestModal/Amount.tsx [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  const [integer, decimal] = amount.split('.')
+  let className = ''
+  if (diffStatus === 'negative') {
+    className = styles.decrease
+  } else if (diffStatus === 'positive') {
+    className = styles.increase
+  }
+
+  return (
+    <div className={className}>
```
[the remark this answers is not in the document]
Comment (yanguoyu) [reply src/components/TransactionItem/TransactionRGBPPDigestModal/Amount.tsx:22]: How about use `styles[diffStatus]`?

File: src/pages/Address/AddressAssetComp.tsx (hunk 1 above)
Comment (Daryl-L) [reply src/pages/Address/AddressAssetComp.tsx]: done

File: src/pages/Address/AddressAssetComp.tsx (hunk 2 above)
Comment (Daryl-L) [reply src/pages/Address/AddressAssetComp.tsx]: done

File: src/pages/Address/AddressAssetComp.tsx (hunk 3 above)
Comment (Daryl-L) [reply src/pages/Address/AddressAssetComp.tsx]: done

File: src/pages/Address/AddressAssetComp.tsx (hunk 4 above)
Comment (Daryl-L) [reply src/pages/Address/AddressAssetComp.tsx]: done

File: src/pages/Address/AddressAssetComp.tsx (hunk 5 above)
Comment (Daryl-L) [reply src/pages/Address/AddressAssetComp.tsx]: done

File: .eslintrc.js (hunk 6 above)
Comment (Daryl-L) [reply .eslintrc.js:129]: > What's this update for

I met an error before, so I added this. Now I find the error disappeared, I will remove it

File: src/components/TransactionItem/TransactionRGBPPDigestModal/TransactionRGBPPDigestContent.tsx (hunk 8 above)
Comment (Daryl-L) [reply src/components/TransactionItem/TransactionRGBPPDigestModal/TransactionRGBPPDigestContent.tsx]: done

File: src/components/TransactionItem/index.tsx (hunk 10 above)
Comment (Daryl-L) [reply src/components/TransactionItem/index.tsx]: I though that the `addressToScript` could return `false`. I just removed it

**Comment by Daryl-L** [comment]
> image The label of BTC address type is inproper, and the `address` is wrapped unexpectedly

done

**Comment by silySuper** [comment]
1.In https://ckb-explorer-frontend-in-magickbase-repo-git-7cb256-magickbase.vercel.app/transaction/0x2127ee68586f16ec4583f2d927cd2fd2dc241878b83a0e53e2cfe09b6fc3b6fa  page ，when search by BTC TX hash,after copy BTC TX hash and click enter key, it will add 0x default，and can not find right result.
截屏2024-03-20 10 11 53

image

**Comment by yanguoyu** [comment]
> 1.In https://ckb-explorer-frontend-in-magickbase-repo-git-7cb256-magickbase.vercel.app/transaction/0x2127ee68586f16ec4583f2d927cd2fd2dc241878b83a0e53e2cfe09b6fc3b6fa page ，when search by BTC TX hash,after copy BTC TX hash and click enter key, it will add 0x default，and can not find right result. 截屏2024-03-20 10 11 53
> 
> image

You'd better copy the full transaction hash without `...`

**Comment by yanguoyu** [comment]
> @Daryl-L Can we add a copy interaction on the BTC TX id? image

image

It already exists

**Comment by silySuper** [comment]
> > @Daryl-L Can we add a copy interaction on the BTC TX id? image
> 
> image
> It already exists

Yes,when click black area ,it will show copied and copy this hash successfully.

**Comment by silySuper** [comment]
1.When scroll up and down browser，tips can not adjust. @devchenyan 

https://github.com/Magickbase/ckb-explorer-frontend/assets/25971273/f4c70f10-9e12-42b6-92a3-3589934ec4bc

**Pull request merged**